sanic/server.py

```diff
@@ -507,8 +516,26 @@ def serve(host, port, request_handler, error_handler, before_start=None,
         for connection in connections:
             connection.close_if_idle()
 
-        while connections:
+        # Gracefully shutdown timeout.
+        # We should provide graceful_shutdown_timeout,
+        # instead of letting connection hangs forever.
+        # Let's roughly calcucate time.
+        start_shutdown = 0
+        while connections and (start_shutdown < graceful_shutdown_timeout):
             loop.run_until_complete(asyncio.sleep(0.1))
+            start_shutdown = start_shutdown + 0.1
+
+        # Force close non-idle connection after waiting for
+        # graceful_shutdown_timeout
+        coros = []
+        for conn in connections:
+            if hasattr(conn, "websocket") and conn.websocket:
+                coros.append(conn.websocket.close_connection(force=True))
+            else:
+                conn.close()
+
+        _shutdown = asyncio.gather(*coros, loop=loop)
```

Comment:
It doesn't matter. you can have a try. i think next line `loop.run_until_complete(_shutdown)` just finishes immediately.